components/function-controller/design/extend_function_spec_proposal.md

```diff
@@ -0,0 +1,392 @@
+# Extending CRD spec for Serverless v1alpha2
+
+## Summary
+
+The current Serverless API allows for limited configuration of the generated Function's deployment. 
+Currently users can only use ENVs to pass 3rd party service credentials but it doesn't allow for volume mounted secrets (which becomes the industry standard for [service bindings](https://servicebinding.io/application-developer/)).
+Moreover users have no control over the annotations that are applied on the function runtime pod. This excludes functions pods from features enabled by annotations (i.e custom log parsers via `fluentbit.io/parser: my-regex-parser`).
+
+
+
+## Motivation
+
+Give Serverless users the ability to:
+- Configure volume mounted secrets (and config maps) for Function's subresources.
+- Configure labels and annotations for the  Function runtime pod.
+
+### Goals
+
+- Add more flexibility to the Serverless API - enable volume mounted secrets and pod annotations.
+- Organise spec attributes belonging to runtime and build-time configuration (?)
+- Provide a rollout plan to implement this functionality in small increments and avoid the need to roll out a new API version
+
+### Non-Goals
+- Define a new API version specification
+
+## Discussion points
+
+### Runtime, build-time separation
+
+Since Function CR is managing two workloads ( Deployment for runtime and Job for build-time ) we need to separate them in the spec in order to make it clear where do the mounts and annotations belong. 
+
+We can either make a clear separation, i.e:
+
+```yaml 
+spec:
+  source:
+  runtimeSpec:
+    # mounts
+    # resources
+    # envs
+    # metadata
+  buildSpec:
+    # resources
+    # envs (?)
+    # metadata (?)
+```
+
+Alternatively we could promote runtime fields to the root (as the belong to kind: Function) and extract only the build-time fields 
+
+```yaml 
+spec:
+  source:
+  # mounts
+  # resources
+  # envs
+  # metadata
+  buildSpec:
+    # resources
+    # envs (?)
+    # metadata (?)
+```
+
+### Mounts - own structure or k8s inherited
+
+Under the hood the secret mount becomes a volume mount in the runtime pod.
+We could :
+ - expose the k8s volume mount spec  in function spec
+ - cover k8s volume mount spec via oor own [specialised](https://servicebinding.io/application-developer/) spec
```

Comment:
Do you mean to reuse service binding from BTP service operator or create something new? First will be too limiting, and second too confusing.

Reply:
I imagine a simple way to define a reference to a secret and optional mounting point (`SERVICE_BINDING_ROOT` env could be used by default ) to support volume mounted secrets as recommended [here](https://servicebinding.io/)
Something like 
```yaml
spec:
  serviceBindings:
  - source: my-redis-secret
  - mountPath: "/redis-secret" # optional
```

Reply:
discussed. and the outcome reflected in the PR